File: figures/src/r/classes/ablation_summary.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

load_primary_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$variant <- as.character(rows$variant)
  rows$module_group <- as.character(rows$module_group)
  rows$label_variant <- as.character(rows$label_variant)
  rows <- rows[order(rows$display_order), ]
  rows
}

load_secondary_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$variant <- as.character(rows$variant)
  rows$module_group <- as.character(rows$module_group)
  rows$metric <- as.character(rows$metric)
  rows$label_variant <- as.character(rows$label_variant)
  rows <- rows[order(rows$display_order, rows$metric), ]
  rows
}

build_source_data <- function(spec, primary_rows, secondary_rows) {
  mapping <- source_data_mapping(spec)
  write_source_csv(
    file.path(repo_root, mapping$a),
    primary_rows,
    c("variant", "module_group", "display_order", "auroc", "delta_auroc", "label_variant")
  )
  write_source_csv(
    file.path(repo_root, mapping$b),
    secondary_rows,
    c("variant", "module_group", "display_order", "metric", "delta_value", "label_variant")
  )
  list(primary = primary_rows, secondary = secondary_rows)
}

group_colors <- function(theme) {
  c(
    "Full system" = theme$palette$neutral[[1]],
    "Architectural module" = theme$palette$categorical[[1]],
    "Training recipe" = theme$palette$categorical[[3]],
    "Objective design" = theme$palette$categorical[[2]],
    "Auxiliary input" = theme$palette$categorical[[4]]
  )
}

summary_text <- function(primary_rows, secondary_rows) {
  largest_drop <- primary_rows[primary_rows$display_order != 1, ][which.min(primary_rows$delta_auroc[primary_rows$display_order != 1]), , drop = FALSE][1, , drop = FALSE]
  ece_rows <- secondary_rows[secondary_rows$metric == "ECE" & secondary_rows$display_order != 1, ]
  worst_ece <- ece_rows[which.max(ece_rows$delta_value), , drop = FALSE][1, , drop = FALSE]
  sprintf("Largest AUROC loss:\n%s\nWorst ECE increase:\n%s", largest_drop$variant, worst_ece$variant)
}

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)
  plot_theme <- build_theme(theme, resolved_font) +
    ggplot2::theme(
      legend.position = "none",
      plot.margin = ggplot2::margin(8, 26, 10, 8)
    )

  primary_rows <- load_primary_rows(file.path(repo_root, spec$data_inputs[[1]]))
  secondary_rows <- load_secondary_rows(file.path(repo_root, spec$data_inputs[[2]]))
  source_rows <- build_source_data(spec, primary_rows, secondary_rows)
  palette <- group_colors(theme)

  primary_plot_rows <- source_rows$primary[nrow(source_rows$primary):1, ]
  primary_plot_rows$variant <- factor(primary_plot_rows$variant, levels = primary_plot_rows$variant)
  primary_plot_rows$auroc_label <- sprintf("%.3f", primary_plot_rows$auroc)

  plot_a <- ggplot2::ggplot(primary_plot_rows, ggplot2::aes(x = delta_auroc, y = variant, fill = module_group)) +
    ggplot2::geom_vline(xintercept = 0, linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_col(width = 0.72, color = theme$palette$neutral[[1]], linewidth = 0.35, alpha = 0.9) +
    ggplot2::geom_text(
      ggplot2::aes(
        label = auroc_label,
        x = ifelse(delta_auroc < 0, delta_auroc - 0.003, delta_auroc + 0.003),
        hjust = ifelse(delta_auroc < 0, 1, 0)
      ),
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::scale_fill_manual(values = palette) +
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
    ggplot2::labs(title = "AUROC drop vs full model", x = "AUROC drop vs full model", y = NULL) +
    ggplot2::coord_cartesian(xlim = c(min(primary_plot_rows$delta_auroc) - 0.018, 0.02), clip = "off") +
    plot_theme +
    ggplot2::theme(
      axis.line.y = ggplot2::element_blank(),
      axis.ticks.y = ggplot2::element_blank()
    )

  secondary_plot_rows <- source_rows$secondary
  variant_levels <- primary_plot_rows$variant
  secondary_plot_rows$variant <- factor(secondary_plot_rows$variant, levels = levels(variant_levels))
  secondary_plot_rows$metric <- factor(secondary_plot_rows$metric, levels = c("AUPRC", "ECE"))
  secondary_plot_rows$label_value <- sprintf("%+.1f%%", secondary_plot_rows$delta_value * 100)

  plot_b <- ggplot2::ggplot(secondary_plot_rows, ggplot2::aes(x = delta_value, y = variant, color = module_group, shape = metric)) +
    ggplot2::geom_vline(xintercept = 0, linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_segment(
      ggplot2::aes(x = 0, xend = delta_value, yend = variant),
      linewidth = 0.5,
      position = ggplot2::position_dodge(width = 0.45),
      alpha = 0.7
    ) +
    ggplot2::geom_point(size = 2.4, stroke = 0.4, position = ggplot2::position_dodge(width = 0.45)) +
    ggplot2::annotate(
      "text",
      x = max(abs(secondary_plot_rows$delta_value)) + 0.017,
      y = 0.55,
      label = summary_text(source_rows$primary, source_rows$secondary),
      hjust = 1,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::annotate(
      "text",
      x = max(abs(secondary_plot_rows$delta_value)) + 0.015,
      y = 1.2,
      label = "circle = AUPRC",
      hjust = 1,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::annotate(
      "text",
      x = max(abs(secondary_plot_rows$delta_value)) + 0.015,
      y = 1.65,
      label = "triangle = ECE",
      hjust = 1,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::scale_color_manual(values = palette) +
    ggplot2::scale_shape_manual(values = c("AUPRC" = 16, "ECE" = 17)) +
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
    ggplot2::labs(title = "Secondary metric shifts", x = "Secondary metric shift", y = NULL) +
    ggplot2::coord_cartesian(
      xlim = c(-max(abs(secondary_plot_rows$delta_value)) - 0.02, max(abs(secondary_plot_rows$delta_value)) + 0.02),
      clip = "off"
    ) +
    plot_theme +
    ggplot2::theme(
      axis.line.y = ggplot2::element_blank(),
      axis.ticks.y = ggplot2::element_blank()
    )

  figure_plot <- (plot_a + plot_b + patchwork::plot_layout(ncol = 2)) +
    patchwork::plot_annotation(
      title = spec$title,
      tag_levels = "a"
    ) &
    ggplot2::theme(
      plot.title = ggplot2::element_text(
        size = as.numeric(theme$typography$title_font_size_pt),
        family = resolved_font$family,
        face = "bold",
        hjust = 0
      ),
      plot.tag = ggplot2::element_text(
        size = as.numeric(theme$panel_labels$font_size_pt),
        face = theme$panel_labels$font_weight,
        family = resolved_font$family
      )
    )

  list(
    plot = figure_plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  figure_plot <- payload$plot
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(figure_plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(figure_plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(figure_plot)
  grDevices::dev.off()

  outputs <- list(
    svg = sub(paste0("^", repo_root, "/"), "", svg_path),
    pdf = sub(paste0("^", repo_root, "/"), "", pdf_path),
    png = sub(paste0("^", repo_root, "/"), "", png_path)
  )
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "ranked_primary_metric_drop_panel",
      "secondary_metric_shift_panel",
      "group_semantic_encoding",
      "metric_shape_encoding_not_color_only",
      "zero_centered_effect_reference",
      "vector_first_export"
    ),
    list(
      panel_count = 2,
      highlight_label_count = 5,
      annotation_count = 10,
      reference_line_count = 2
    )
  )
  outputs
}

File: figures/src/python/classes/ablation_summary.py
#!/usr/bin/env python3
"""Class-based Python renderer for ablation-summary figures."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    add_panel_labels,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


def _load_primary_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "variant": row["variant"],
                "module_group": row["module_group"],
                "display_order": int(row["display_order"]),
                "auroc": float(row["auroc"]),
                "delta_auroc": float(row["delta_auroc"]),
                "label_variant": row["label_variant"],
            }
            for row in reader
        ]


def _load_secondary_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "variant": row["variant"],
                "module_group": row["module_group"],
                "display_order": int(row["display_order"]),
                "metric": row["metric"],
                "delta_value": float(row["delta_value"]),
                "label_variant": row["label_variant"],
            }
            for row in reader
        ]


def build_source_data(
    spec: dict[str, Any], primary_rows: list[dict[str, Any]], secondary_rows: list[dict[str, Any]]
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    ordered_primary = sorted(primary_rows, key=lambda item: item["display_order"])
    ordered_secondary = sorted(
        secondary_rows, key=lambda item: (item["display_order"], item["metric"])
    )
    mapping = source_data_mapping(spec)
    export_csv(
        REPO_ROOT / mapping["a"],
        ordered_primary,
        ["variant", "module_group", "display_order", "auroc", "delta_auroc", "label_variant"],
    )
    export_csv(
        REPO_ROOT / mapping["b"],
        ordered_secondary,
        ["variant", "module_group", "display_order", "metric", "delta_value", "label_variant"],
    )
    return ordered_primary, ordered_secondary


def _group_colors(theme: dict[str, Any]) -> dict[str, str]:
    return {
        "Full system": theme["palette"]["neutral"][0],
        "Architectural module": theme["palette"]["categorical"][0],
        "Training recipe": theme["palette"]["categorical"][2],
        "Objective design": theme["palette"]["categorical"][1],
        "Auxiliary input": theme["palette"]["categorical"][3],
    }


def _metric_styles() -> dict[str, dict[str, Any]]:
    return {
        "AUPRC": {"marker": "o", "offset": -0.15},
        "ECE": {"marker": "s", "offset": 0.15},
    }


def _summary_text(primary_rows: list[dict[str, Any]], secondary_rows: list[dict[str, Any]]) -> str:
    largest_drop = min(
        (row for row in primary_rows if row["display_order"] != 1),
        key=lambda item: item["delta_auroc"],
    )
    worst_ece = max(
        (row for row in secondary_rows if row["metric"] == "ECE" and row["display_order"] != 1),
        key=lambda item: item["delta_value"],
    )
    return (
        f"Largest AUROC loss:\n{largest_drop['variant']}\n"
        f"Worst ECE increase:\n{worst_ece['variant']}"
    )


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    primary_rows = _load_primary_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    secondary_rows = _load_secondary_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    panel_a_rows, panel_b_rows = build_source_data(spec, primary_rows, secondary_rows)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt
    from matplotlib.ticker import PercentFormatter

    fig, axes = plt.subplots(
        1,
        2,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        gridspec_kw={"width_ratios": [1.0, 1.0]},
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)

    colors = _group_colors(theme)
    metric_styles = _metric_styles()

    panel_a = axes[0]
    ordered_primary = list(reversed(panel_a_rows))
    y_positions = list(range(len(ordered_primary)))
    values = [float(row["delta_auroc"]) for row in ordered_primary]
    panel_a.barh(
        y_positions,
        values,
        color=[colors[str(row["module_group"])] for row in ordered_primary],
        edgecolor=theme["palette"]["neutral"][0],
        linewidth=0.4,
        alpha=0.9,
    )
    panel_a.axvline(0.0, color=theme["palette"]["neutral"][1], linewidth=0.9)
    panel_a.set_yticks(y_positions)
    panel_a.set_yticklabels([str(row["variant"]) for row in ordered_primary])
    panel_a.set_xlabel("AUROC drop vs full model")
    panel_a.set_title("AUROC drop vs full model", loc="left", pad=3)
    panel_a.xaxis.set_major_formatter(PercentFormatter(xmax=1.0, decimals=0))
    panel_a.set_xlim(min(values) - 0.018, 0.02)
    panel_a.spines["top"].set_visible(False)
    panel_a.spines["right"].set_visible(False)
    panel_a.spines["left"].set_visible(False)
    panel_a.grid(axis="x", color=theme["palette"]["neutral"][2], linewidth=0.6)
    for y_pos, row in zip(y_positions, ordered_primary):
        value = float(row["delta_auroc"])
        panel_a.text(
            value - 0.003 if value < 0 else value + 0.003,
            y_pos,
            f"{float(row['auroc']):.3f}",
            ha="right" if value < 0 else "left",
            va="center",
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
        )

    panel_b = axes[1]
    variants = list(reversed(panel_a_rows))
    base_y = {str(row["variant"]): index for index, row in enumerate(variants)}
    panel_b.axvline(0.0, color=theme["palette"]["neutral"][1], linewidth=0.9)
    for row in panel_b_rows:
        metric = str(row["metric"])
        style = metric_styles[metric]
        y_value = base_y[str(row["variant"])] + float(style["offset"])
        x_value = float(row["delta_value"])
        panel_b.plot(
            [0.0, x_value],
            [y_value, y_value],
            color=colors[str(row["module_group"])],
            linewidth=1.1,
            alpha=0.75,
        )
        panel_b.scatter(
            [x_value],
            [y_value],
            color=colors[str(row["module_group"])],
            marker=style["marker"],
            s=34,
            edgecolor=theme["palette"]["neutral"][0],
            linewidth=0.35,
            zorder=3,
        )
    panel_b.set_yticks(list(base_y.values()))
    panel_b.set_yticklabels(list(base_y.keys()))
    panel_b.set_xlabel("Secondary metric shift")
    panel_b.set_title("Secondary metric shifts", loc="left", pad=3)
    panel_b.xaxis.set_major_formatter(PercentFormatter(xmax=1.0, decimals=0))
    extent = max(abs(float(row["delta_value"])) for row in panel_b_rows) + 0.02
    panel_b.set_xlim(-extent, extent)
    panel_b.spines["top"].set_visible(False)
    panel_b.spines["right"].set_visible(False)
    panel_b.spines["left"].set_visible(False)
    panel_b.grid(axis="x", color=theme["palette"]["neutral"][2], linewidth=0.6)
    panel_b.text(
        1.02,
        0.96,
        "circle = AUPRC",
        transform=panel_b.transAxes,
        ha="left",
        va="top",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
    )
    panel_b.text(
        1.02,
        0.88,
        "square = ECE",
        transform=panel_b.transAxes,
        ha="left",
        va="top",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
    )
    panel_b.text(
        0.98,
        0.03,
        _summary_text(panel_a_rows, panel_b_rows),
        transform=panel_b.transAxes,
        ha="right",
        va="bottom",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
    )

    add_panel_labels(spec, theme, [panel_a, panel_b])
    fig.text(
        0.01,
        0.965,
        spec["title"],
        ha="left",
        va="top",
        fontsize=theme["typography"]["title_font_size_pt"],
        fontweight="bold",
        color=theme["palette"]["neutral"][0],
    )
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)
    fig = create_figure(spec_path)

    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = spec["figure_id"]
    outputs = {
        "svg": str((output_dir / f"{stem}.svg").relative_to(REPO_ROOT)),
        "pdf": str((output_dir / f"{stem}.pdf").relative_to(REPO_ROOT)),
        "png": str((output_dir / f"{stem}.png").relative_to(REPO_ROOT)),
    }
    fig.savefig(REPO_ROOT / outputs["svg"], format="svg", transparent=False)
    fig.savefig(REPO_ROOT / outputs["pdf"], format="pdf", transparent=False)
    fig.savefig(
        REPO_ROOT / outputs["png"],
        format="png",
        dpi=int(profile["preview_dpi"]),
        transparent=False,
    )

    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "ranked_primary_metric_drop_panel",
            "secondary_metric_shift_panel",
            "group_semantic_encoding",
            "metric_shape_encoding_not_color_only",
            "zero_centered_effect_reference",
            "vector_first_export",
        ],
        {
            "panel_count": 2,
            "highlight_label_count": 5,
            "annotation_count": 10,
            "reference_line_count": 2,
        },
    )

    import matplotlib.pyplot as plt

    plt.close(fig)
    return outputs


if __name__ == "__main__":
    build_figure(Path("figures/specs/figure_11_ablation_summary.yml"))
